core/src/main/java/org/apache/brooklyn/core/location/access/PublicNetworkFaceEnricher.java

```diff
@@ -0,0 +1,373 @@
+            return transformPort(entity, machine, port);
+        } else if (isUri(sensorVal)) {
+            return transformUri(entity, machine, sensorVal.toString());
+        } else if (isHostAndPort(sensorVal)) {
+            return transformHostAndPort(entity, machine, sensorVal.toString());
+        } else {
+            // no-op; unrecognised type
+            return Maybe.absent();
+        }
+    }
+
+    protected boolean isUri(Object sensorVal) {
+        if (sensorVal instanceof URI || sensorVal instanceof URL) {
+            return true;
+        }
+        try {
+            new URI(sensorVal.toString());
+            return true;
+        } catch (URISyntaxException e) {
+            return false;
+        }
+    }
+
+    protected boolean isPort(Object sensorVal) {
+        return (sensorVal instanceof Integer || sensorVal instanceof Long || sensorVal instanceof Short);
+    }
+
+    protected int toInteger(Object sensorVal) {
+        if (sensorVal instanceof Number) {
+            return ((Number)sensorVal).intValue();
+        } else {
+            throw new IllegalArgumentException("Expected number but got "+sensorVal+" of type "+(sensorVal != null ? sensorVal.getClass() : null));
+        }
+    }
+
+    protected boolean isHostAndPort(Object sensorVal) {
+        if (sensorVal instanceof HostAndPort) {
+            return true;
+        } else if (sensorVal instanceof String) {
+            try {
+                HostAndPort hostAndPort = HostAndPort.fromString((String)sensorVal);
+                return hostAndPort.hasPort();
+            } catch (IllegalArgumentException e) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    protected Maybe<String> transformUri(Entity source, MachineLocation machine, String sensorVal) {
+        URI uri = URI.create(sensorVal);
+        int port = uri.getPort();
+        if (port == -1 && "http".equalsIgnoreCase(uri.getScheme())) port = 80;
+        if (port == -1 && "https".equalsIgnoreCase(uri.getScheme())) port = 443;
+
+        if (port != -1) {
+            HostAndPort publicTarget = getPortForwardManager().lookup(machine, port);
+            if (publicTarget == null) {
+                // TODO What if publicTarget is still null, but will be set soon? We're not subscribed to changes in the PortForwardManager!
+                // TODO Should we return null or sensorVal? In this method we always return sensorVal;
+                //      but in HostAndPortTransformingEnricher we always return null!
+                LOG.trace("network-facing enricher not transforming {} URI {}, because no port-mapping for {}", new Object[] {source, sensorVal, machine});
+                return Maybe.absent();
+            }
+            URI result;
+            try {
+                result = new URI(uri.getScheme(), uri.getUserInfo(), publicTarget.getHostText(), publicTarget.getPort(), uri.getPath(), uri.getQuery(), uri.getFragment());
```

Comment:
For the case where no explicit port is defined in source url - It's quite possible that the port mapping has the same public port as the internal one. Suggest comparing and passing -1 for the port if that's the case so that the URL doesn't contain default ports.